### PR TITLE
Descriptive error messages for missing template hooks errors

### PR DESCRIPTION
**Pull request by Vild**
This is my warmup project for GSoC that was suggested by my mentors. This PR implemented more descriptive error messages for missing template hooks. This replaces following the none-descriptive error message: `Error: undefined identifier __equals in module object` with:
> Error: object.__equals not found. The current runtime does not support equal checks on arrays, or the runtime is corrupt.

The error message is modeled after ObjectNotFound: https://github.com/dlang/dmd/blob/8e364d300adef422a36ac39b0f09dcfaed78312a/src/dmd/declaration.d#L202

Questions:
- Does verifyHookExist use the best way of checking for an identifier inside the `object` module?
- `ObjectNotFound` have the "or corrupt" part, should this error message also have it?
- Any improvements to make the error message more clear?

---

Output from this PR with a empty object.d and -checkaction=context:
```d
struct MyStruct { int a, b; }
MyStruct[] castToMyStruct(int[] arr) {
	return cast(MyStruct[])arr;	// Error: object.__ArrayCast not found. The current runtime does not support casting array of structs, or the runtime is corrupt.
}

extern(C) int main(int argc, char** argv) {
	int[] arrA, arrB;
	bool a = arrA[] == arrB[]; // Error: object.__equals not found. The current runtime does not support equal checks on arrays, or the runtime is corrupt.
	bool b = arrA < arrB; // Error: object.__cmp not found. The current runtime does not support comparing arrays, or the runtime is corrupt.

	{
		int x = 1; int y = 1;
		assert(x == y); // Error: object._d_assert_fail not found. The current runtime does not support generating assert messages, or the runtime is corrupt.
	}

	switch ("a") { // Error: object.__switch not found. The current runtime does not support switch cases on strings, or the runtime is corrupt.
	default:
		break;
	}

	final switch (0) { // Error: object.__switch_error not found. The current runtime does not support generating assert messages, or the runtime is corrupt.
	case 1:
		break;
	}

	return 0;
}

```

### CODE REVIEW

**Comment by wilzbach** [comment]
Nice! Could you add test cases for this, s.t. the new code is covered? Have a look at `test/run.d` and the readme in test for details

**Comment by JinShil** [comment]
> std/json.d(702): Error: `object.__equals` not found. The current runtime does not support equal checks on arrays, or the runtime is corrupt.

That's strange.  Is it a false positive?

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -11746,3 +11772,25 @@ VarDeclaration makeThis2Argument(const ref Loc loc, Scope* sc, FuncDeclaration f
...
+ * Params:
+ *      loc = location to use for error messages
+ *      sc = current scope
+ *      id = the hook identifier
+ *      description = what the hook does
+ *      module_ = what module the hook is located in
+ * Returns:
+ *      If the hook was found.
+ */
+extern (C++) bool verifyHookExist(const ref Loc loc, Scope* sc, Identifier id, const(char)* description, Identifier module_ = Id.object)
```
Comment (JinShil) [inline src/dmd/expressionsem.d]: Favor `ref Scope sc` if possible.

**Comment by JinShil** [comment]
> Does verifyHookExist use the best way of checking for an identifier inside the object module?

File: src/dmd/expressionsem.d (hunk 1 above)
Comment (JinShil) [reply src/dmd/expressionsem.d]: Also, favor `string` instead of `const(char)*`.  We're trying to get rid of `char*` in DMD.

File: src/dmd/expressionsem.d [hunk 2]
```diff
@@ -10653,10 +10673,16 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+            {
+                result = new ErrorExp();
+                return;
+            }
+
             Expression __equals = new IdentifierExp(exp.loc, Id.empty);
-            Identifier id = Identifier.idPool("__equals");
             __equals = new DotIdExp(exp.loc, __equals, Id.object);
-            __equals = new DotIdExp(exp.loc, __equals, id);
+            __equals = new DotIdExp(exp.loc, __equals, Id.__equals);
```
Comment (JinShil) [inline src/dmd/expressionsem.d]: It's best to put things like this in a separate PR.

File: src/dmd/expressionsem.d (hunk 1 above)
Comment (Vild) [reply src/dmd/expressionsem.d]: > Favor `ref Scope sc` if possible.

Scope is stored as a pointer inside ExpressionSemanticVisitor.
I could change it to a ref, but then I need to dereference it each time `verifyHookExist` is called
Should I still do it?

> Also, favor `string` instead of `const(char)*`. We're trying to get rid of `char*` in DMD.

The `const(char)*` is sent all the way down to the C function `vsnprintf` via `OutBuffer`, so it need to be a `const(char)*`. I can't find a `tostringz` function so I could not got that way either.

File: src/dmd/expressionsem.d (hunk 2 above)
Comment (Vild) [reply src/dmd/expressionsem.d]: Will do

**Comment by Vild** [comment]
> > std/json.d(702): Error: `object.__equals` not found. The current runtime does not support equal checks on arrays, or the runtime is corrupt.
> 
> That's strange. Is it a false positive?

The problem here is that when I search for `Id.object` it finds `std.json.JSONValue.object`.
I guess I need to find another way of referring directly to the `object` module.

**Comment by wilzbach** [comment]
> I guess I need to find another way

Yes, but should be able to start with object and then use a dotTemplate instantiation for its members.

File: src/dmd/expressionsem.d (hunk 1 above)
Comment (JinShil) [reply src/dmd/expressionsem.d]: > Should I still do it?

Comment (JinShil) [reply src/dmd/expressionsem.d]: > I can't find a tostringz function so I could not got that way either.

Is it possible for you to use `f("%.*s", cast(int)s.length, s.ptr)` to workaround the current limitations of using `string` in DMD?  That technique is used in several other places in DMD currently.

Comment (jacob-carlborg) [reply src/dmd/expressionsem.d]: Yes, `"%.*s"` will work.

Comment (jacob-carlborg) [reply src/dmd/expressionsem.d]: > I can't find a tostringz function so I could not got that way either.

There's one here:

https://github.com/dlang/dmd/blob/df6418f0dd4cfeb79129a717f1a1037ccf4c90ae/src/dmd/utils.d#L207

**Comment by Vild** [comment]
I've implemented fixed for everything so far:
- I had to start with `Id.empty` and then go to `object`, to correctly find `object.d`
- I removed the small refactoring that JinShin said
- verifyHookExist now take `Scope` as a ref, and `description` as a string.
  - I used `%.*s` to print the string, as suggested.
- I added a small failure test-case that verifies that all the error messages are triggered and outputted correctly.

File: src/dmd/expressionsem.d [hunk 3]
```diff
@@ -11748,3 +11764,28 @@ VarDeclaration makeThis2Argument(const ref Loc loc, Scope* sc, FuncDeclaration f
...
+ *      id = the hook identifier
+ *      description = what the hook does
+ *      module_ = what module the hook is located in
+ * Returns:
+ *      If the hook was found.
+ */
+bool verifyHookExist(const ref Loc loc, ref Scope sc, Identifier id, string description, Identifier module_ = Id.object)
+{
+    Dsymbol rootSymbol = sc.search(loc, Id.empty, null);
+    Dsymbol moduleSymbol = rootSymbol.search(loc, module_);
```
Comment (jacob-carlborg) [inline src/dmd/expressionsem.d]: * This could be moved into the if-statement
* `auto` could be used

File: src/dmd/expressionsem.d [hunk 4]
```diff
@@ -11748,3 +11764,28 @@ VarDeclaration makeThis2Argument(const ref Loc loc, Scope* sc, FuncDeclaration f
...
+    if (moduleSymbol)
+    {
+      Dsymbol idSymbol = moduleSymbol.search(loc, id);
+      if (idSymbol)
```
Comment (jacob-carlborg) [inline src/dmd/expressionsem.d]: Indentation looks wrong.

File: test/fail_compilation/verifyhookexist.d [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+fail_compilation/verifyhookexist.d(102): Error: `object.__ArrayCast` not found. The current runtime does not support casting array of structs, or the runtime is corrupt.
+fail_compilation/verifyhookexist.d(200): Error: `object.__equals` not found. The current runtime does not support equal checks on arrays, or the runtime is corrupt.
+fail_compilation/verifyhookexist.d(201): Error: `object.__cmp` not found. The current runtime does not support comparing arrays, or the runtime is corrupt.
+fail_compilation/verifyhookexist.d(205): Error: `object._d_assert_fail` not found. The current runtime does not support generating assert messages, or the runtime is corrupt.
+fail_compilation/verifyhookexist.d(208): Error: `object.__switch` not found. The current runtime does not support switch cases on strings, or the runtime is corrupt.
+fail_compilation/verifyhookexist.d(213): Error: `object.__switch_error` not found. The current runtime does not support generating assert messages, or the runtime is corrupt.
+---
+*/
+
+#line 100
```
Comment (jacob-carlborg) [inline test/fail_compilation/verifyhookexist.d]: Please avoid using these `#line` tokens. There's a command to automatically update the error messages: https://github.com/dlang/dmd/blob/master/test/README.md#automatically-update-the-test_output-segments

File: src/dmd/expressionsem.d (hunk 3 above)
Comment (Vild) [reply src/dmd/expressionsem.d]: Better like this?
```d
auto rootSymbol = sc.search(loc, Id.empty, null);
if (auto moduleSymbol = rootSymbol.search(loc, module_))
    if (moduleSymbol.search(loc, id))
        return true;
```

Comment (jacob-carlborg) [reply src/dmd/expressionsem.d]: Possibly, I'm not sure.

**Review by JinShil — APPROVED** [review]

File: src/dmd/expressionsem.d [hunk 6]
```diff
@@ -5718,6 +5718,10 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                 tok == TOK.in_ ||
                 isEqualsCallExpression)
             {
+                // Verify that the hook `object._d_assert_fail` exist
```
Comment (Basile-z) [inline src/dmd/expressionsem.d]: The comment is useless because the function self documents enough

File: src/dmd/expressionsem.d [hunk 7]
```diff
@@ -11748,3 +11764,24 @@ VarDeclaration makeThis2Argument(const ref Loc loc, Scope* sc, FuncDeclaration f
...
+/*******************************
+ * Make sure that the runtime hook exists.
+ * Params:
+ *      loc = location to use for error messages
+ *      sc = current scope
+ *      id = the hook identifier
+ *      description = what the hook does
+ *      module_ = what module the hook is located in
+ * Returns:
+ *      If the hook was found.
```
Comment (Basile-z) [inline src/dmd/expressionsem.d]:
``` 
`true` if ... otherwise `false` 
```

or maybe

```
a `bool` indicating if the hook is present
```

File: src/dmd/expressionsem.d [hunk 8]
```diff
@@ -11748,3 +11764,24 @@ VarDeclaration makeThis2Argument(const ref Loc loc, Scope* sc, FuncDeclaration f
     vthis2.nestedrefs.push(fd);
     return vthis2;
 }
+
+/*******************************
+ * Make sure that the runtime hook exists.
```
Comment (Basile-z) [inline src/dmd/expressionsem.d]: `a runtime hook` rather than `the`

**Review by ghost — APPROVED** [review]
The check is obviously OK but the associated comments need a few corrections IMO. Also in the error message is mentioned a possible corruption of the runtime. Can this really happen ?

File: src/dmd/expressionsem.d (hunk 6 above)
Comment (Basile-z) [reply src/dmd/expressionsem.d]: ditto for the other similar code comments placed before a call to the check.

**Comment by Vild** [comment]
> Also in the error message is mentioned a possible corruption of the runtime. Can this really happen ?

I based it on the error that happens if TypeInfo is missing: `TypeInfo not found. object.d may be incorrectly installed or corrupt.`.

Another reason for keeping this part is in case the error get triggered, and the user does not really know what a runtime is, it is probably better to say that the runtime could be corrupted instead of just saying that the current runtime does not support feature *X*.

I guess it could be changed to something like `The current runtime does not support [...], or runtime version mismatch.`, but I think this is less clear than the corrupt part.

File: src/dmd/expressionsem.d (hunk 8 above)
Comment (JinShil) [reply src/dmd/expressionsem.d]: I disagree with that.  "Make sure a runtime hook exists" means you are checking for the existence of at least 1 runtime hook.  "Make sure the runtime hook exists" means you are checking for the existence of a specific runtime hook.

Comment (Basile-z) [reply src/dmd/expressionsem.d]: I meant that the function is not specilized for a particular hook, i.e "the hook" it checks "a hook" specified with a parameter.

Comment (Vild) [reply src/dmd/expressionsem.d]: I'm changing it to `Make sure that the runtime hook $(PARAM id) exists.` to specify that it looks for the hook `id`.

Comment (jacob-carlborg) [reply src/dmd/expressionsem.d]: Or `Make sure that the given runtime hook exists`. Now it can be confusing that it looks explicitly for a hook named "id".

**Comment by Basile-z** [comment]
lgtm

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
@Vild this good to go?

**Comment by Vild** [comment]
> @Vild this good to go?

Yup, I'm happy with the PR.
